Handle flow extra errors

cideM opened the pull request:
Flow now reports errors in a slightly different format, which also includes an `extra` field. See the below JSON, from the issue #900 I created.

```
"errors": [
    {
      "extra": [
        {
          "message": [
            {
              "context": "null",
              "descr": "Property `handleOnSubmit` is incompatible:",
              "type": "Blame",
              "path": "",
              "line": 0,
              "endline": 0,
              "start": 1,
              "end": 0
            }
          ],
          "children": [
            {
              "message": [
                {
                  "context": "handleOnSubmit = async () => {",
                  "descr": "Promise",
                  "type": "Blame",
                  "loc": {
                    "source": "/home/yuuki/tiniestfinance/src/App.js",
                    "type": "SourceFile",
                    "start": {
                      "line": 73,
                      "column": 31,
                      "offset": 1787
                    },
                    "end": {
                      "line": 73,
                      "column": 31,
                      "offset": 1788
                    }
                  },
                  "path": "/home/yuuki/tiniestfinance/src/App.js",
                  "line": 73,
                  "endline": 73,
                  "start": 31,
                  "end": 31
                },
                {
                  "context": "null",
                  "descr": "This type is incompatible with",
                  "type": "Comment",
                  "path": "",
                  "line": 0,
                  "endline": 0,
                  "start": 1,
                  "end": 0
                },
                {
                  "context": "  handleOnSubmit: (event?: Event) => void,",
                  "descr": "undefined",
                  "type": "Blame",
                  "loc": {
                    "source": "/home/yuuki/tiniestfinance/src/App.js",
                    "type": "SourceFile",
                    "start": {
                      "line": 14,
                      "column": 38,
                      "offset": 356
                    },
                    "end": {
                      "line": 14,
                      "column": 41,
                      "offset": 360
                    }
                  },
                  "path": "/home/yuuki/tiniestfinance/src/App.js",
                  "line": 14,
                  "endline": 14,
                  "start": 38,
                  "end": 41
                }
              ]
            }
          ]
        }
      ]
```

I added a block to `flow.vim` which is heavily inspired by the existing code. I also added one flag `g:ale_javascript_flow_extra` and added documentation for the feature. I did not write tests as the idea behind the handler tests isn't quite clear to me. But I may also be overlooking something here, as the files required in the `test_flow_handler.vader` file all seem to point at local files which are not under source control and therefore won't work with CI. I'd happily write tests for the added feature if someone could point me in the right direction.

Also, these are the first vimL lines I've written so if there are mistakes, let me know and I'll try to fix them.

The end result is that instead of just
```
src/App.js|89 col 17 error| props of React element `New`: This type is incompatible with object type
```
you would now also get 
```
src/App.js|89 col 17 error| Property `setVector` is incompatible: number This type is incompatible with function type
```

w0rp commented:
I wonder if it's worth turning this on by default. It seems pretty useful.

w0rp commented:
It doesn't seem like it could do any harm, other than just telling you too much about what went wrong.

You might consider displaying one error at the cursor instead of two. If there are two, then the displaying of the errors at the cursor won't work so well. You could maybe put the message on the end of the string, or put it in the `detail` key. Then you'd be able to see the explanation for why something went wrong with `:ALEDetail`.

mvolkmann commented:
Agreed!

On Thu, Sep 21, 2017 at 4:43 PM, w0rp <notifications@github.com> wrote:

> I wonder if it's worth turning this on by default. It seems pretty useful.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/w0rp/ale/pull/946#issuecomment-331290530>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAE10IXd6KuxRtwiFaJn5Ekn2AdnoviXks5sktiRgaJpZM4Pf86Z>
> .
>



-- 
R. Mark Volkmann
Object Computing, Inc.

cideM commented:
I'm on holidays right now and don't have any time for programming but I'll implement the suggestions next week!

w0rp commented:
Sounds good. :+1:

cideM commented:
I moved the extra messages to detail. Do you still want to keep the flag then? As you already mentioned, it would now probably make a lot of sense to just have this enabled per default, as it's hidden behind `:ALEDetail` anyway.

w0rp commented:
I think you can get rid of the flag, and just always include the extra details. Make sure to update the tests appropriately.

cideM commented:
It's been a while... I really need to work on priorities. Damn you Overwatch.

Anyway, I added a Vader test for extra errors. As usually, I derived some significant inspiration from existing tests. Everything was green on my machine/docker. Also removed the flag and all traces of it. @w0rp

w0rp commented on ale_linters/javascript/flow.vim:
```diff
@@ -71,36 +74,70 @@ function! ale_linters#javascript#flow#Handle(buffer, lines) abort
-        endfor
+        let l:detail = ''
+
+        if has_key(l:error, 'message')
```
What happens when the error doesn't have this key?

w0rp commented on ale_linters/javascript/flow.vim:
```diff
@@ -71,36 +74,70 @@ function! ale_linters#javascript#flow#Handle(buffer, lines) abort
+                    let l:text = l:message.descr . ':'
+                else
+                    let l:text = l:text . ' ' . l:message.descr
+                endif
```
This function has become pretty long. This whole block could be moved to another function.

w0rp commented on ale_linters/javascript/flow.vim:
```diff
@@ -71,36 +74,70 @@ function! ale_linters#javascript#flow#Handle(buffer, lines) abort
             let l:text = l:text . ' See also: ' . l:error.operation.descr
         endif
 
+        if has_key(l:error, 'extra')
```
Could you move this part to another function and call it here? You could call it `s:GetDetail` or something similar.

w0rp commented on ale_linters/javascript/flow.vim:
```diff
@@ -71,36 +74,70 @@ function! ale_linters#javascript#flow#Handle(buffer, lines) abort
         \})
-    endfor
+
+      endfor
```
The indentation went a bit weird here.

w0rp commented on ale_linters/javascript/flow.vim:
```diff
@@ -71,36 +74,70 @@ function! ale_linters#javascript#flow#Handle(buffer, lines) abort
         \   'text': l:text,
-        \   'type': l:error.level is# 'error' ? 'E' : 'W',
+        \   'type': has_key(l:error, 'level') && l:error.level is# 'error' ? 'E' : 'W',
+        \   'detail': l:detail,
```
Be aware that `:ALEDetail` will show nothing if an empty string is given here.

cideM commented:
I extracted two functions: one for crafting the detail message and one for handling the extra errors generally. Also changed the way the final object for `call` is crafted I hope that is in some way idiomatic vimscript. This should hopefully avoid sending the `detail` key along if it just has an empty string in it.

w0rp commented:
Cool, looks good. :+1: Thank you for all of your work on this.

w0rp commented:
Cheers! :beers:

cideM commented:
Yay! It was fun 🎊
